amespahdbpythonsuite/parsers.py

```diff
@@ -0,0 +1,543 @@
+        """
+
+        if self.check_schema:
+            tree = ElementTree.parse(self.filename)
+            root = tree.getroot()
+            self._validate_schema(tree, root)
+            formatted_tree = \
+                ElementTree.iterwalk(tree, events=("start", "end"))
+
+        else:
+            formatted_tree = \
+                ElementTree.iterparse(self.filename, events=("start", "end"))
+
+        database, info = self._tree_to_dict(formatted_tree)
+
+        return database, info
+
+    def _validate_schema(self, tree, root):
+        """Validate XML schema.
+
+        Args:
+            tree (ElementTree.ElementTree): parsed XML file.
+            root: root element of ElementTree tree.
+
+        Returns:
+            N/A. If the XML schema is successfully validated, the instance
+            variable self.validated is set to True (bool).
+        """
+
+        schema = root.get('{http://www.w3.org/2001/XMLSchema-instance}' +
+                          'schemaLocation')
+        if schema:
+            _, uri = schema.split(' ', 1)
+            doc = ElementTree.parse(urllib.request.urlopen(uri))
+            xmlschema = ElementTree.XMLSchema(doc)
+            try:
+                xmlschema.assertValid(tree)
+            except Exception as e:
+                raise e
+            else:
+                self.validated = True
+
+    def _tree_to_dict(self, tree):
+        """Convert the tree to a dict of dicts.
+
+        Args:
+            tree: lxml tree of PAHdb file.
+
+        Returns:
+            database_dict: Dictionary, with the UIDs as keys,
+                containing the transitions, geometry data, as well as
+                UID metadata, references, comments.
+            info_dict: Dictionary containing general information about
+                the XML file, including filename, type, date, full
+                (bool), version, comment, and number of species.
+        """
+        # Advance to root and identify namespace.
+        event, root = next(tree)
+        namespace = self.determine_xml_namespace(root)
+
+        # Root dict attributes.
+        root_dict = root.attrib
+
+        # Form important dicts, to be returned.
+        database_dict = {}
+        info_dict = {}
+        first_run = True
+
+        # Iterate through the tree, clearing elements as we go.
+        for _, (event, elem) in enumerate(tree):
+
+            # Store the top XML comment.
+            if first_run:
+                if event == "end" and elem.tag == namespace + 'comment':
+                    pahdb_comment = elem.text
+                    first_run = False
+
+            # Extract UID data on a per-tag ('specie') basis.
+            if event == "end" and elem.tag == namespace + 'specie':
+                # Extract soft data.
+                uid, metadata, comments, reference = \
+                    self._extract_info(elem, namespace)
+
+                # Extract hard data.
```

Comment:
What is 'hard' data?

Reply:
The numeric data -- updated the descriptions to be more accurate.